### ISSUE TITLE
Syntax Highlighting and Custom Styles

### ISSUE THREAD

**Issue by xsc** [weavejester/codox#135]
I've been working on making codox' capable of providing syntax highlighting (using [highlight.js](https://highlightjs.org/)), as well as of using different stylesheets. The work can be found in [this repository](https://github.com/xsc/codox-themes).

There is unfortunately no real diff since I split up functionality into multiple namespaces but here's what I did:

- use [markdown-clj](https://github.com/yogthos/markdown-clj) for Markdown rendering to generate output compatible with highlight.js,
- include highlight.js assets (base JS + desired theme's CSS),
- make main stylesheet configurable.

I don't think it makes sense to have a separate project for this, so – if desired – I'd open a PR to include the changes (without the namespace split) into codox itself. (And it shouldn't be a large changeset.)

This would result in the following enhancements:

- support for customizable syntax highlighting,
- support for different styles for the generated documentation,
- might address things like #71, #109.

Tell me what you think!

**Comment by weavejester** [comment]
This looks good. Thanks for your work on it!

Codox currently uses Pegdown, but I've been meaning to switch to a Clojure version as Pegdown has some longstanding bugs and work seems to have stalled on it. Unfortunately markdown-clj doesn't support some of Pegdown's additional features, such as definition lists, that are particularly useful for Codox. Switching to markdown-clj at the moment would result in some functionality being lost, and I'd like to avoid that if possible. So using highlight.js means either adding the missing functionality to markdown-clj, or find a way to use highlight.js with Pegdown. I personally favor the former option, if it's at all possible.

Codox does provide HTML transformations which can be used to change the stylesheet or to add a new one, but it might be useful to add a simpler option to address this case. How are you handling adding themes? Perhaps a theme could be just be a bunch of resources plus an edn transformation file in a predefined location.

**Comment by xsc** [comment]
> So using highlight.js means either adding the missing functionality to markdown-clj, or find a way to use highlight.js with Pegdown. I personally favor the former option, if it's at all possible.

It might be easier to make Pegdown render fenced code blocks like markdown-clj than to add a multitude of unrelated features to the latter – but I'll look into it!

>  How are you handling adding themes?

Currently, themes are CSS only, i.e. there are some CSS files on the classpath which can be referenced by name and replace `default.css`. The actual HTML is not touched.

> Perhaps a theme could be just be a bunch of resources plus an edn transformation file in a predefined location.

Sounds very good. For example, we could map the theme `:awesome` to a Clojure namespace `codox.theme.awesome` which contains a function with a well-known name, e.g. `theme`, producing a map with all the data we need for rendering:

```clojure
(defn theme
  [project]
  (-> {:css {"css/default.css" (io/resource "codox-awesome/base.css")
           "css/fontawesome.css" (io/resource "codox-awesome/fontawesome.css")}
       :writer codox.theme/default-writer
       :transformations ...}
      (codox.theme/include-highlight-js :solarized-light)
      (codox.theme/include-jquery)))
```

(OT: Ideally, the writer itself would be split up further into `:head` (rendering contents of the `<head>...</head>` element), `:layout` (rendering the static parts of the page, like the sidebar and header), `:index`, `:namespace`, `:document`. This would give us some more flexibility/composability – but it's also a significant and breaking change regarding the writer.)

This way, users can add themes as `:dependencies` in the project.clj, specify something like `:codox {:theme :awesome}` and be done.

**Comment by weavejester** [comment]
I believe Pegdown does have support for [custom code formatters](https://github.com/sirthias/pegdown/issues/20), so it might be easiest to add it to Pegdown. Eventually I'd like to migrate away from Pegdown though, since development on it seems to have stalled, and it has a few annoying bugs.

Regarding the theme, should it be a function, or just an edn file that represents a collection of HTML transformations? For example:

```edn
[[:head] [:append [:link {:rel "stylesheet", :type "text/css" :href "theme/awesome/base.css"}]
                  [:script {:src "theme/awesome/highlight.js"}]]
```

There could be a structure with the following resources:

* `codox/theme/awesome/transforms.edn`
* `codox/theme/awesome/public/base.css`
* `codox/theme/awesome/public/highlight.js`

Anything in `public` gets added to the output directory, and transforms.edn contains the transformations to apply to the HTML.

If the user wants something more sophisticated they can use a custom writer, but I'd imagine that most themes would consist CSS, JavaScript and perhaps some static images.

**Comment by xsc** [comment]
While I agree that any sophisticated changes to the output should probably end up in a separate writer, I'm not entirely sure I'd want to come up with a serialisation format for enlive transformations, as well as the respective runtime. I think I'd prefer to have the transformations directly as enlive data structures.

Some questions:

- Would we still include `highlight.js` in codox itself? Or would themes provide the necessary transformations to the code blocks, as well as any assets themselves? (Might produce some duplication between themes – but I don't think that'd be a significant problem.)
- Would we keep the default assets (`default.css`, jQuery, ...) and let themes override them or would every theme be self-contained?

**Comment by weavejester** [comment]
> I'm not entirely sure I'd want to come up with a serialisation format for enlive transformations, as well as the respective runtime.

This functionality is already in Codox. See [HTML Transformations](https://github.com/weavejester/codox#html-transformations) in the README, and the `codox.writer.html/transform-html` function.

> Would we still include `highlight.js` in codox itself?

Yep, I think highlighting code with highlight.js is a good default. The only reason I haven't done it yet is because I haven't had time, but it's been on my todo list for a while.

> Would we keep the default assets and let themes override them or would every theme be self-contained?

Just make every theme self-contained and use the HTML transformations to add or remove HTML code as needed.

**Comment by xsc** [comment]
If it's okay for you, I'd add [cpath-clj](https://github.com/xsc/cpath-clj) for resource collection (since we want to collect everything within a certain directory on the classpath and that directory will most likely be within a JAR).

Apart from that, I'd now try to implement your suggestion, i.e. copy all public resources directly to the output path and apply the transformations to add `<script>`, `<link>`, ... tags.

**Comment by weavejester** [comment]
Fixed by #138.
